File: python/OOP/linked list.py
class Node:
    
        
    def __init__(self, value):
        self.value = value
        self.next = None

            
class LinkedList:
    
    
    def __init__(self):
        self.head = None
    
    def addToEnd(self, value):
        new_node = Node(value)
        if self.head is None:
            self.head = new_node 
        else:
            islast = self.head
            while islast.next:
                islast = islast.next
            islast.next = new_node
            
    def insert(self, value, pos):
        new_node = Node(value)
        if self.head is None:
            self.head = new_node 
            print('Списк был пуст. Элемент добавлен в начало')
        elif pos == 0:
            new_node.next = self.head
            self.head = new_node 
        elif self.head is not None:
            i = 1
            prev = self.head          
            while prev.next:       
                if i == pos:                
                    new_node.next = prev.next
                    prev.next = new_node    
                    break
                i += 1
                prev = prev.next     
            else:
                self.addToEnd(value)
            
    def public(self):
        islast = self.head
        while islast.next:
            print(islast.value.value, end=", ")
            islast = islast.next
        print(islast.value.value)
        
    def checkElem(self, value):
        islast = self.head.value
        i = 0
        while islast:
            if value == islast:
                print(f'Элемент {value.value} имеет индекс {i}')
                break
            else:
                islast = islast.next
                i += 1
                
    def removeElem(self, pos):
        node = self.head
        if pos == 0:
            self.head = node.next
            node = None
        else:
            i = 1
            while node.next:
                prev_node = node
                node = node.next
                if pos == i:
                    prev_node.next = node.next
                    node = None
                    break
                else:
                    i += 1
                    prev_node = node.next
            

a = Node(12)
b = Node(56)
c = Node(87)
d = Node(109)
e = Node(333)
g = Node(333)
ll = LinkedList()
ll.addToEnd(a)
ll.addToEnd(b)
ll.addToEnd(c)
ll.addToEnd(d)
print('список')
ll.public()
ll.insert(e, 7)
print('после вставки')
ll.public()
ll.removeElem(2)
print('после удаления')
ll.public()
ll.checkElem(a)

File: java/LinkedList/src/linkedlist/LinkedList.java
package linkedlist;
public class LinkedList {
    
    Node head = null;
    
    public void insert(Node val) {
        
        Node new_node = val;
        if(this.head == null) {
            this.head = new_node;
        } else {
            Node islast = this.head;
            while(islast.next != null) {
                islast = islast.next;
                }
            islast.next = new_node;
        }
    }
    
    public void insert(Node val, int pos) {
        
        Node new_node = val;
        Node node = this.head;
        int i = 1;
        int size = this.getSize();
        if (0 > pos | pos > size) {
            System.out.println("Insert position out of list size");
        } else if (pos == 0) {
            new_node.next = node;
            this.head = new_node;
        } else {
            while (i != pos) {
                node = node.next;
                i = i + 1;
            } 
            new_node.next = node.next;
            node.next = new_node;
        }
    }  
    
    public void remove(int pos) {
        Node node = this.head;
        int i = 1;
        int size = this.getSize();
        if (pos < 0 | pos >= size) {
            System.out.println("Remove position out of list lenght");
        } else if (pos == 0){
            System.out.println("Removing element - " + node.val);
            this.head = node.next;
        } else {
            while(i != pos) {
                node = node.next;
                i++;
                } 
            System.out.println("Removing element - " + node.next.val);
            node.next = node.next.next;
            }
    }
    
    public void getElem(int pos){
        int i = 0;
        Node node = this.head;
        int size = this.getSize();
        if (0 > pos | pos > size) {
            System.out.println("Search position out of list size");
        } else { while(i != pos) {
                    node = node.next;
                    i++;
                }
            System.out.println(node.val);
            }
    }
    
    public int getSize() {
        Node islast = this.head;
        int i = 1;
        while (islast.next != null) {
            islast = islast.next;
            i++;
        }
        return i;
    }
    
    public void getList() {
        Node first = this.head;
        while(first.next != null) {
            System.out.print(first.val + ", ");
            first = first.next; 
        }
        System.out.println(first.val);
    }

    public static void main(String[] args) {
        LinkedList ll = new LinkedList();
        Node a = new Node(7); 
        Node b = new Node(14); 
        Node c = new Node(73); 
        Node d = new Node(56);
        Node g = new Node(2456);
        ll.insert(a);
        ll.insert(b);
        ll.insert(c);
        ll.insert(d);
        ll.insert(g,5);
        System.out.println("List size - " + ll.getSize());
        ll.getList();
        ll.getElem(12);
        ll.remove(4);
        ll.getList();
    }
}